runtime/near-vm-runner/src/cache.rs

```diff
@@ -85,6 +87,110 @@ impl fmt::Debug for MockCompiledContractCache {
     }
 }
 
+struct FilesystemCompiledContractCacheState {
+    dir: rustix::fd::OwnedFd,
+    test_temp_dir: Option<tempfile::TempDir>,
+}
+
+#[derive(Clone)]
+pub struct FilesystemCompiledContractCache {
+    state: Arc<FilesystemCompiledContractCacheState>,
+}
+
+impl FilesystemCompiledContractCache {
+    pub fn new<HD: AsRef<std::path::Path> + ?Sized, SP: AsRef<std::path::Path> + ?Sized>(
+        home_dir: &HD,
+        store_path: Option<&SP>,
+    ) -> std::io::Result<Self> {
+        let path = std::env::var_os("NEAR_COMPILED_CACHE");
+        let path = path.map(PathBuf::from).unwrap_or_else(|| {
+            let store_path = store_path.map(AsRef::as_ref).unwrap_or_else(|| "data".as_ref());
+            home_dir.as_ref().join(store_path).join("contracts")
+        });
+        std::fs::create_dir_all(&path)?;
+        let dir = rustix::fs::open(path, rustix::fs::OFlags::DIRECTORY, rustix::fs::Mode::empty())?;
+        Ok(Self {
+            state: Arc::new(FilesystemCompiledContractCacheState { dir, test_temp_dir: None }),
+        })
+    }
+
+    pub fn test() -> std::io::Result<Self> {
+        let tempdir = tempfile::TempDir::new()?;
+        let mut cache = Self::new(tempdir.path(), None::<&str>)?;
+        Arc::get_mut(&mut cache.state).unwrap().test_temp_dir = Some(tempdir);
+        Ok(cache)
+    }
+}
+
+const ERROR_TAG: u8 = b'\n';
+const CODE_TAG: u8 = 0b10010101;
```

Comment:
How did you choose those tags? 
nit: Just by looking at it I can't tell for sure that those are different bytes :)